credentialz/README.md

```diff
@@ -341,6 +376,126 @@ stream.Send(
 )
 ```
 
+## User Journeys
+
+### Rotate Certificate based on existing key
+
+The most common operation we are expecting to require on devices is the rotation of certificates used for SSH access for devices. This operation expects to reuse the existing host key on the device.
+
+* Get the public key configured on the host.
+
+```go
+
+resp, err := c.GetPublicKeys(&GetPublicKeyRequests{})
+```
+
+* Generate certificate based on key.
+
+* Rotate certificate on device.
+
+```go
+stream.Send(
+    RotateHostCredentialsRequest {
+        server_keys: ServerKeysRequest {
+            certificate: "A....=",
+            version: "v1.0",
+            created_on: 3214451134,
+        }
+    }
+)
+```
+
+* Validate that new settings are working as expected.
+
+* Finalize request.
+
+```go
+stream.Send(
+    RotateHostCredentialsResponse {
+        finalize: FinalizeRequest {}
+    }
+)
+```
+
+### Generate new host key on device and rotate certificate based on the new key
+
+This use case focuses on the rotation of a host key and then generation of the certificate based on the new public key.
+
+* Send request for generation of new private key.
+
+```go
+stream.Send(
+    RotateHostCredentialsRequest {
+        generate_keys: []GenerateKeysRequest {{
+            key_params: KeyGen.KEY_GEN_SSH_KEY_TYPE_EDDSA_ED25519 
+        }}
+    }
+)
+```
+
+* Get Response containing public key to generate the certificate.
+
+```go
+resp, err := stream.Recv()
+data := resp.PublicKeys
+```
+
+* The caller will then use this data to generate a certificate.
+
+* Send generated cert to device to rotate.
+
+```go
+stream.Send(
+    RotateHostCredentialsRequest {
+        server_keys: ServerKeysRequest {
+            certificate: "A....=",
+            version: "v1.0",
+            created_on: 3214451134,
+        }
+    }
+)
+```
+
+* Validate the `RotateCredentialsResponse`.
+
+```go
+if _, err := stream.Recv(); err != nil {
+    ...
+}
+```
+
+* (Proposal) As there would be a state after the generate key step but before the certficate is returned, the device will not be able to have alignment between the previous cert and new key. We are proposing to add a new state which allows for the atomic rotation of both the key and certificate by holding the rotation of the key until a `Rotate` message is sent after the Response message.
+
+```go
+stream.Send(
+    RotateHostCredentialsRequest {
```

Comment:
no it is not optional, the whole point of adding this step is to address your desire to have this as a way to lock step the state of the key and cert.

Reply:
When we brought this up before we raised it as a concern and you seemed fine with that. To be clear - Arista is fine with having a process where:
- At the start a hostkey / ssh cert are configured
- An RPC connection is opened
- A new hostkey is configured and immediately put into place. The old certificate goes away.
- A new certificate is configured and immediately put into place.
- A finalize command is issued
- The RPC is closed.

When we left the meeting, we thought that was your desired behavior as well. It doesn't quite lead to nice atomic behavior since you can't generate a key on-box and have a new certificate as one operation, but it's not a big deal as long as everyone involved is okay with it. Our goal was to make sure that keys generated on the switch are possible to use for all the flows since the original flow was "keys sent over the wire"-centric.

The problem with having a mandatory RotateIdentity message is that now:
- Do we need this for every flow? If we do then this does feel like a major API change because all the clients need to follow this new flow.
- If we don't then the gNSI server is trying to figure out what flow it is and decide where we are in the sequence. The RPC command is now more stateful than before.

Reply:
good then i will delete it literally the only reason i put it in as a proposal was based on your feedback in the meeting